core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala

```diff
@@ -0,0 +1,145 @@
+   *
+   * @param field the BQ schema's field definition
+   * @param row all field values for a single BQ row
+   * @return the Rawls attributename->attribute pair, post-translation
+   */
+  def fieldToAttribute(field: Field, row: FieldValueList): (AttributeName, Attribute) = {
+    val attrName = field.getName
+    val fv = Try(row.get(attrName)) match {
+      case Success(fieldValue) => fieldValue
+      case Failure(ex) =>
+        throw new DataEntityException(s"field ${field.getName} not found in row ${asDelimitedString(row)}")
+    }
+
+    val attribute = field.getMode match {
+      case Mode.REPEATED =>
+        fv.getRepeatedValue.asScala.toList match {
+          case x if x.isEmpty => AttributeValueEmptyList
+          case members =>
+            // can a list itself contain a list? Recursion here is difficult because of the Attribute/AttributeValue hierarchy
+            val attrValues = members.map { member => fieldValueToAttributeValue(field, member) }
+            AttributeValueList(attrValues)
+        }
+      case _ => fieldValueToAttributeValue(field, fv)
+    }
+
+    (AttributeName.withDefaultNS(attrName), attribute)
+  }
+
+  /**
+   * Translates a BigQuery result set into a list of Rawls Entities. Returns an empty list if BigQuery returned zero rows.
+   *
+   * @param queryResults the BigQuery result set
+   * @param entityType name of a BigQuery table in the result set, used as the Rawls entity type.
+   * @param primaryKey what Rawls believes is the identifying column in the BigQuery table, used to generate entity names.
+   * @return list of Rawls Entities
+   */
+  def queryResultsToEntities(queryResults:TableResult, entityType: String, primaryKey: String): List[Entity] = {
+    // short-circuit if query results is empty
+    if (queryResults.getTotalRows == 0) {
+      List.empty[Entity]
+    } else {
+
+      val schemaFields = queryResults.getSchema.getFields
+
+      // does primary key exist in the results?
+      if (Try(schemaFields.getIndex(primaryKey)).isFailure) {
+        throw new DataEntityException(s"could not find primary key column '$primaryKey' in query results: ${asDelimitedString(schemaFields)}")
+      }
+
+      // short-circuit if query results is empty
+      if (queryResults.getTotalRows == 0) {
+        List.empty[Entity]
+      } else {
+        val fieldDefs:List[Field] = schemaFields.iterator().asScala.toList
+        queryResults.iterateAll().asScala.map { row =>
+          val attrs = fieldDefs.map { field => fieldToAttribute(field, row) }.toMap
+          val entityName = attrs.find(_._1.name == primaryKey) match {
+            case Some((_, a: Attribute)) => AttributeStringifier.apply(a)
+            case None =>
+              // this shouldn't happen, since we validated the pk against the schema
+              throw new DataEntityException(s"could not find primary key column '$primaryKey' on record: ${asDelimitedString(schemaFields)}")
```

Comment:
This exception is the same as the one on line 94 - can we define it once and reuse?

Reply:
they're subtly different, and their wordings are subtly different. On line 94, we're checking to see if the primary key from the method arguments exists in the BigQuery schema - in other words, does the table have the requested column? That happens before looking at any of the rows, as a quick validation step.

On line 108, we're looping through and checking each individual row to see if that row contains the requested column. This should always pass, since if it's in the schema, each record should have it.

Reply:
Since this error is row-specific, then we should probably replace `${asDelimitedString(schemaFields)}` with `${asDelimitedString(row)}` so that users know which row is missing the primary key

Reply:
good call, updated!